Repair action & resource bundle

MaanooAk opened the pull request:
Repair of units based on the [doc/reverse_engineering/repair.md](doc/reverse_engineering/repair.md) (file in pull request).

In order to continue:
* ~~A data structure for multiple resource costs should be implemented (game_resource_bundle ?)~~
* The cost of the unit should be stored somewhere (inside unt_type ?)

~~Can I implement these? or just leave as it is and revisit after the nyan revolution?~~

EDIT: Also game_resource_bundle should replace the `game_resource type, int value` couples all over the place.

MaanooAk commented:
ResourceBoundle could be like [this (gist)](https://gist.github.com/MaanooAk/58cded6e8ac2fac9a48f66eda0630aba). Check the [main](https://gist.github.com/MaanooAk/58cded6e8ac2fac9a48f66eda0630aba#file-resource-h-L95) for how its used.

MaanooAk commented:
Implemented a ResourceBundle class and used it in player (kept old one with raw type and amount) and unit repair.

Next step is to add somewhere (unit type, unit attributes, ?) the cost of the unit. I have zero confidence implementing that so **this is as far I can go** (= removing WIP).

TheJJ commented:
Sorry for the delay, university has kept me very busy (http://www.move2space.de/MOVE-II/)...

However, I just tested your code and I couldn't get my villagers to repair their town center or mill.

The resource bundle is a good idea, as it's capable of adding resource types at runtime easily. We'll need that in the future.

The cost of the unit is drawn from the gamedata of course, which is accessible through the `std::vector<gamedata::empiresdat> gamedata;` in `libopenage/gamestate/game_spec.cpp`. But this is all going to be replaced by nyan. But nyan is still quite far from being ready so adding just another dirty hardcoding for it is ok. You'll find the structure of that struct in `openage/convert/gamedata` where we extract it.

MaanooAk commented:
I added the repair ability, that should do the work. I added and some other missing abilities while I was there.
And I implemented the repair gui button.

TheJJ commented on libopenage/unit/ability.cpp:
```diff
@@ -270,6 +275,125 @@ void AttackAbility::invoke(Unit &to_modify, const Command &cmd, bool play_sound)
+}
+
+void ConvertAbility::invoke(Unit &to_modify, const Command &cmd, bool play_sound) {
+	to_modify.log(MSG(dbg) << "invoke repair action");
```
copypasta :)

TheJJ commented on libopenage/unit/ability.h:
```diff
@@ -264,6 +265,103 @@ class AttackAbility: public UnitAbility {
 
+/*
+ * initiates a repair action when given a valid target
+ */
```
pls write `/**` so it's parsed by doxygen

TheJJ commented on libopenage/unit/ability.h:
```diff
@@ -264,6 +265,103 @@ class AttackAbility: public UnitAbility {
+
+/*
+ * initiates a heal action when given a valid target
+ */
```
same here

TheJJ commented on libopenage/unit/ability.h:
```diff
@@ -264,6 +265,103 @@ class AttackAbility: public UnitAbility {
+ * initiates a research
+ * TODO implement
+ */
+class ResearchAbility: public UnitAbility {
```
and here

TheJJ commented on libopenage/unit/ability.h:
```diff
@@ -264,6 +265,103 @@ class AttackAbility: public UnitAbility {
+ * initiates a patrol action when given a valid target
+ * TODO implement
+ */
+class PatrolAbility: public UnitAbility {
```
and here

TheJJ commented on libopenage/unit/ability.h:
```diff
@@ -264,6 +265,103 @@ class AttackAbility: public UnitAbility {
+/*
+ * initiates a convert action when given a valid target
+ */
+class ConvertAbility: public UnitAbility {
```
and here

TheJJ commented on libopenage/unit/action.h:
```diff
@@ -419,6 +420,11 @@ class RepairAction: public TargetAction {
+	float time;
+	float time_left;
+
+	ResourceBundle cost; // calculate and store the cost in the constructor
```
you can write `/** stores the cost of the repair */` or whatever is more suitable above it so it's parsed by doxygen, too.

MaanooAk commented:
The separate namespace should be done with a refactoring tool.
However I think the way attributes are defined is nice too.

TheJJ commented:
You mean templating the attributes? Yes, could also be done :)
Do whatever you like better and think suits us more.

Shall I merge your stuff before that or do you wanna integrate it also in this pull?

MaanooAk commented:
You should merge this (do you want me to rebase?)
I'm thinking of fixing the attributes first (more like #651) if that's ok.

TheJJ commented:
No I don't think you need to rebase, it doesn't conflict with the `master`.
@ attributes, sure go ahead! It's another basic foundation for the game logic.

MaanooAk commented:
I just realized I'm creating a ton of conflicts in my new pr (#683) with this one...
Can you merge this or you are not sure yet, can I do something?

TheJJ commented on libopenage/unit/ability.cpp:
```diff
@@ -270,6 +275,125 @@ void AttackAbility::invoke(Unit &to_modify, const Command &cmd, bool play_sound)
+}
+
+void HealAbility::invoke(Unit &to_modify, const Command &cmd, bool play_sound) {
+	to_modify.log(MSG(dbg) << "invoke repair action");
```
still the wrong message here

MaanooAk commented:
I see... Just leave it, I will investigate and come back to you.
(May take some time, I have some serious Ubuntu craziness happening right now)

MaanooAk commented:
Added the ability, added 1000 of each res to players (so they can repair) and made gaia the Gaia player (prev: name1 was Gaia and gaia was some civ)

![screenshot from 2016-12-30 22-45-34](https://cloud.githubusercontent.com/assets/6997990/21572532/9759fc0a-cee2-11e6-9552-232c79c6d3c5.png)

Yanikore commented:
I'd like to add to this that there's a bug in the original game which still exists where you can repair (and garrison) a building from a diagonal location. It's called the wall corner glitch and I'm wondering if this bug also exists within openage and if it should either be fixed, or stay exactly the same.

TheJJ commented:
@Yanikore I think we'll have the same problem right now because we don't have any blocking physics currently, the pathfinder is really stupid and doesn't account for that. Go and help out at #366 :)

Yanikore commented:
@TheJJ I'd love to help out but I'm not capable for the C++/python side of this. I'm more a PHP developer than anything so I'll probably help out in the future when a website gets made. Right now I can only add some conceptual ideas to the situation.

TheJJ commented:
yay it works :)
